// prettier-plugin-svelte/src/embed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVisitorKeys = getVisitorKeys;
exports.embed = embed;
const prettier_1 = require("prettier");
const getText_1 = require("./lib/getText");
const snipTagContent_1 = require("./lib/snipTagContent");
const options_1 = require("./options");
const doc_helpers_1 = require("./print/doc-helpers");
const helpers_1 = require("./print/helpers");
const node_helpers_1 = require("./print/node-helpers");
const extractAttributes_1 = require("./lib/extractAttributes");
const base64_string_1 = require("./base64-string");
const { builders: { group, hardline, softline, indent, dedent, literalline }, utils: { removeLines }, } = prettier_1.doc;
const leaveAlone = new Set([
    'Script',
    'Style',
    'Identifier',
    'MemberExpression',
    'CallExpression',
    'ArrowFunctionExpression',
]);
const dontTraverse = new Set(['start', 'end', 'type']);
function getVisitorKeys(node, nonTraversableKeys) {
    return Object.keys(node).filter((key) => {
        return !nonTraversableKeys.has(key) && !leaveAlone.has(node.type) && !dontTraverse.has(key);
    });
}
// Embed works like this in Prettier v3:
// - do depth first traversal of all node properties
// - deepest property is calling embed first
// - if embed returns a function, it will be called after the traversal in a second pass, in the same order (deepest first)
// For performance reasons we try to only return functions when we're sure we need to transform something.
function embed(path, _options) {
    const node = path.getNode();
    const options = _options;
    if (!options.locStart || !options.locEnd || !options.originalText) {
        throw new Error('Missing required options');
    }
    if ((0, helpers_1.isASTNode)(node)) {
        (0, node_helpers_1.assignCommentsToNodes)(node);
        attachAttributeComments(node);
        if (node.module) {
            node.module.type = 'Script';
            node.module.attributes = (0, extractAttributes_1.extractAttributes)((0, getText_1.getText)(node.module, options));
        }
        if (node.instance) {
            node.instance.type = 'Script';
            node.instance.attributes = (0, extractAttributes_1.extractAttributes)((0, getText_1.getText)(node.instance, options));
        }
        if (node.css) {
            node.css.type = 'Style';
            // Preserve the snipped style-body marker on top-level <style> blocks so embedTag()
            // can round-trip the real CSS body in our fixture package path.
            node.css.attributes = (0, extractAttributes_1.extractAttributes)((0, getText_1.getText)(node.css, options));
            node.css.content.type = 'StyleProgram';
        }
        return null;
    }
    // embed does depth first traversal with deepest node called first, therefore we need to
    // check the parent to see if we are inside an expression that should be embedded.
    const parent = path.getParentNode();
    const printJsExpression = () => parent.expression ? printJS(parent, 'expression', {}) : undefined;
    const printSvelteBlockJS = (name) => printJS(parent, name, { forceSingleLine: true });
    switch (parent.type) {
        case 'IfBlock':
        case 'ElseBlock':
        case 'AwaitBlock':
        case 'KeyBlock':
            printSvelteBlockJS('expression');
            break;
        case 'EachBlock':
            printSvelteBlockJS('expression');
            printSvelteBlockJS('key');
            break;
        case 'SnippetBlock':
            // We merge the two parts into one expression to then treat it like a function
            if (node === parent.expression) {
                parent.expression.end =
                    options.originalText.indexOf(')', parent.parameters?.[parent.parameters.length - 1]?.typeAnnotation
                        ?.end ??
                        parent.parameters?.[parent.parameters.length - 1]?.end ??
                        parent.expression.end) + 1;
                parent.parameters = null;
                node.isJS = true;
                node.asFunction = true;
            }
            break;
        case 'Element':
            printJS(parent, 'tag', {});
            break;
        case 'MustacheTag':
            printJS(parent, 'expression', {
                forceSingleQuote: (0, node_helpers_1.isInsideQuotedAttribute)(path, options),
            });
            break;
        case 'RawMustacheTag':
            printJS(parent, 'expression', {});
            break;
        case 'Spread':
            printJS(parent, 'expression', {});
            break;
        case 'AttachTag':
            printJS(parent, 'expression', {});
            break;
        case 'ConstTag':
            printJS(parent, 'expression', { removeParentheses: true });
            break;
        case 'Binding':
            printJS(parent, 'expression', {
                removeParentheses: parent.expression.type === 'SequenceExpression',
                surroundWithSoftline: true,
            });
            break;
        case 'RenderTag':
            if (node === parent.expression) {
                // TODO: remove this if block at some point, snippet API changed in .next-..
                if ('argument' in parent || 'arguments' in parent) {
                    parent.expression.end =
                        options.originalText.indexOf(')', parent.argument?.end ??
                            parent.arguments?.[parent.arguments.length - 1]?.end ??
                            parent.expression.end) + 1;
                    parent.argument = null;
                    parent.arguments = null;
                }
                printJS(parent, 'expression', {});
            }
            break;
        case 'EventHandler':
        case 'Binding':
        case 'Class':
        case 'Let':
        case 'Transition':
        case 'Action':
        case 'Animation':
        case 'InlineComponent':
            printJsExpression();
            break;
    }
    if (node.isJS) {
        return async (textToDoc) => {
            try {
                const embeddedOptions = {
                    // Prettier only allows string references as parsers from v3 onwards,
                    // so we need to have another public parser and defer to that
                    parser: options._svelte_ts
                        ? 'svelteTSExpressionParser'
                        : 'svelteExpressionParser',
                    singleQuote: node.forceSingleQuote ? true : options.singleQuote,
                    _svelte_asFunction: node.asFunction,
                };
                // If we have snipped content, it was done wrongly and we need to unsnip it.
                // This happens for example for {@html `<script>{foo}</script>`}
                const text = (0, getText_1.getText)(node, options, true);
                let docs = await textToDoc(node.asFunction ? forceIntoFunction(text) : forceIntoExpression(text), embeddedOptions);
                if (node.forceSingleLine) {
                    docs = removeLines(docs);
                }
                if (node.removeParentheses) {
                    docs = (0, doc_helpers_1.removeParentheses)(docs);
                }
                if (node.asFunction) {
                    if (Array.isArray(docs) && typeof docs[0] === 'string') {
                        docs[0] = docs[0].replace('function ', '');
                        docs.splice(-1, 1);
                    }
                    else {
                        throw new Error('Prettier AST changed, asFunction logic needs to change');
                    }
                }
                if (node.surroundWithSoftline) {
                    docs = group(indent([softline, group(docs), dedent(softline)]));
                }
                return docs;
            }
            catch (e) {
                return (0, getText_1.getText)(node, options, true);
            }
        };
    }
    const embedType = (tag, parser, isTopLevel) => {
        return async (textToDoc, print) => {
            return embedTag(tag, options.originalText, path, (content) => formatBodyContent(content, parser, textToDoc, options), print, isTopLevel, options);
        };
    };
    const embedScript = (isTopLevel) => embedType('script', 
    // Use babel-ts as fallback because the absence does not mean the content is not TS,
    // the user could have set the default language. babel-ts will format things a little
    // bit different though, especially preserving parentheses around dot notation which
    // fixes https://github.com/sveltejs/prettier-plugin-svelte/issues/218
    (0, node_helpers_1.isTypeScript)(node) ? 'typescript' : (0, node_helpers_1.isJSON)(node) ? 'json' : 'babel-ts', isTopLevel);
    const embedStyle = (isTopLevel) => embedType('style', (0, node_helpers_1.isLess)(node) ? 'less' : (0, node_helpers_1.isScss)(node) ? 'scss' : 'css', isTopLevel);
    const embedPug = () => embedType('template', 'pug', false);
    switch (node.type) {
        case 'Script':
            return embedScript(true);
        case 'Style':
            return embedStyle(true);
        case 'Element': {
            if (node.name === 'script') {
                return embedScript(false);
            }
            else if (node.name === 'style') {
                return embedStyle(false);
            }
            else if ((0, node_helpers_1.isPugTemplate)(node)) {
                return embedPug();
            }
        }
    }
    return null;
}
function forceIntoExpression(statement) {
    // note the trailing newline: if the statement ends in a // comment,
    // we can't add the closing bracket right afterwards
    return `(${statement}\n)`;
}
function forceIntoFunction(statement) {
    return `function ${statement} {}`;
}
function preformattedBody(str) {
    if (!str) {
        return '';
    }
    const firstNewline = /^[\t\f\r ]*\n/;
    const lastNewline = /\n[\t\f\r ]*$/;
    // If we do not start with a new line prettier might try to break the opening tag
    // to keep it together with the string. Use a literal line to skip indentation.
    return [literalline, str.replace(firstNewline, '').replace(lastNewline, ''), hardline];
}
function getSnippedContent(node, options) {
    const encodedContent = (0, node_helpers_1.getAttributeTextValue)(snipTagContent_1.snippedTagContentAttribute, node);
    if (encodedContent) {
        return (0, base64_string_1.base64ToString)(encodedContent);
    }
    // Some top-level Svelte AST nodes, especially <style>, may drop the synthetic snipped-content
    // attribute from their normalized `attributes` array even though the marker is still present in
    // `options.originalText`. Fall back to slicing the raw tag text and unsnipping that source span.
    const rawTagText = (0, getText_1.getText)(node, options, true);
    const match = rawTagText.match(/^<\w+\b[^>]*>([\s\S]*)<\/\w+>$/i);
    return match ? match[1] : '';
}
async function formatBodyContent(content, parser, textToDoc, options) {
    try {
        const body = await textToDoc(content, { parser });
        if (parser === 'pug' && typeof body === 'string') {
            // Pug returns no docs but a final string.
            // Therefore prepend the line offsets
            const whitespace = options.useTabs
                ? '\t'
                : ' '.repeat(options.pugTabWidth && options.pugTabWidth > 0
                    ? options.pugTabWidth
                    : options.tabWidth);
            const pugBody = body
                .split('\n')
                .map((line) => (line ? whitespace + line : line))
                .join('\n');
            return [hardline, pugBody];
        }
        const indentIfDesired = (doc) => options.svelteIndentScriptAndStyle ? indent(doc) : doc;
        (0, doc_helpers_1.trimRight)([body], doc_helpers_1.isLine);
        return [indentIfDesired([hardline, body]), hardline];
    }
    catch (error) {
        if (process.env.PRETTIER_DEBUG) {
            throw error;
        }
        // We will wind up here if there is a syntax error in the embedded code. If we throw an error,
        // prettier will try to print the node with the printer. That will fail with a hard-to-interpret
        // error message (e.g. "Unsupported node type", referring to `<script>`).
        // Therefore, fall back on just returning the unformatted text.
        console.error(error);
        return preformattedBody(content);
    }
}
async function embedTag(tag, text, path, formatBodyContent, print, isTopLevel, options) {
    const node = path.getNode();
    const content = tag === 'template' ? (0, node_helpers_1.printRaw)(node, text) : getSnippedContent(node, options);
    const previousComments = node.type === 'Script' || node.type === 'Style'
        ? node.comments
        : [(0, node_helpers_1.getLeadingComment)(path)]
            .filter(Boolean)
            .map((comment) => ({ comment: comment, emptyLineAfter: false }));
    const canFormat = (0, node_helpers_1.isNodeSupportedLanguage)(node) &&
        !(0, node_helpers_1.isIgnoreDirective)(previousComments[previousComments.length - 1]?.comment) &&
        (tag !== 'template' ||
            options.plugins.some((plugin) => typeof plugin !== 'string' && plugin.parsers && plugin.parsers.pug));
    const body = canFormat
        ? content.trim() !== ''
            ? await formatBodyContent(content)
            : content === ''
                ? ''
                : hardline
        : preformattedBody(content);
    const openingTag = group([
        '<',
        tag,
        indent(group([
            ...path.map((0, helpers_1.printWithPrependedAttributeLine)(node, options, print), 'attributes'),
            (0, options_1.isBracketSameLine)(options) ? '' : dedent(softline),
        ])),
        '>',
    ]);
    let result = group([openingTag, body, '</', tag, '>']);
    const comments = [];
    for (const comment of previousComments) {
        comments.push('<!--', comment.comment.data, '-->');
        comments.push(hardline);
        if (comment.emptyLineAfter) {
            comments.push(hardline);
        }
    }
    if (isTopLevel && options.svelteSortOrder !== 'none') {
        // top level embedded nodes have been moved from their normal position in the
        // node tree. if there is a comment referring to it, it must be recreated at
        // the new position.
        return [...comments, result, hardline];
    }
    else {
        // Only comments at the top level get the special "move comment" treatment.
        return isTopLevel && comments.length ? [...comments, result] : result;
    }
}
function printJS(node, name, options) {
    const part = node[name];
    if (!part || typeof part !== 'object') {
        return;
    }
    part.isJS = true;
    part.forceSingleQuote = options.forceSingleQuote;
    part.forceSingleLine = options.forceSingleLine;
    part.removeParentheses = options.removeParentheses;
    part.surroundWithSoftline = options.surroundWithSoftline;
}
/**
 * Walk the AST and use `_comments` (stashed by the parser) to attach
 * attribute-level comments to their neighbouring attribute nodes via
 * Prettier's `util.addLeadingComment` / `util.addTrailingComment`.
 */
function attachAttributeComments(ast) {
    const comments = ast._comments;
    if (!comments || comments.length === 0)
        return;
    // Index comments by start position for fast lookup
    const commentsByStart = new Map();
    for (const c of comments) {
        commentsByStart.set(c.start, c);
    }
    walkAndAttach(ast.html, commentsByStart);
}
function walkAndAttach(node, commentsByStart) {
    if (!node || typeof node !== 'object')
        return;
    if ('attributes' in node && Array.isArray(node.attributes) && node.attributes.length > 0) {
        const attrs = node.attributes;
        // Check gap before first attribute (between tag name and first attr)
        const tagNameEnd = node.start + 2;
        attachCommentsInRange(tagNameEnd, attrs[0].start, null, attrs[0], commentsByStart);
        // Check gaps between consecutive attributes
        for (let i = 0; i < attrs.length - 1; i++) {
            attachCommentsInRange(attrs[i].end, attrs[i + 1].start, attrs[i], attrs[i + 1], commentsByStart);
        }
    }
    // Recurse into children and block branches
    for (const child of (0, node_helpers_1.getChildren)(node)) {
        walkAndAttach(child, commentsByStart);
    }
    if ((node.type === 'IfBlock' || node.type === 'EachBlock') && node.else) {
        walkAndAttach(node.else, commentsByStart);
    }
    if (node.type === 'AwaitBlock') {
        if (node.pending)
            walkAndAttach(node.pending, commentsByStart);
        if (node.then)
            walkAndAttach(node.then, commentsByStart);
        if (node.catch)
            walkAndAttach(node.catch, commentsByStart);
    }
}
function attachCommentsInRange(rangeStart, rangeEnd, precedingAttr, followingAttr, commentsByStart) {
    for (const [start, comment] of commentsByStart) {
        if (start >= rangeStart && comment.end <= rangeEnd) {
            if (followingAttr) {
                prettier_1.util.addLeadingComment(followingAttr, comment);
            }
            else if (precedingAttr) {
                prettier_1.util.addTrailingComment(precedingAttr, comment);
            }
            commentsByStart.delete(start);
        }
    }
}
