// prettier-plugin-svelte/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.options = exports.printers = exports.parsers = exports.languages = void 0;
const prettierPluginBabel = require("prettier/plugins/babel");
const print_1 = require("./print");
const embed_1 = require("./embed");
const snipTagContent_1 = require("./lib/snipTagContent");
const compiler_1 = require("svelte/compiler");
const babelParser = prettierPluginBabel.parsers.babel;
const typescriptParser = prettierPluginBabel.parsers['babel-ts']; // TODO use TypeScript parser in next major?
function locStart(node) {
    return node.start;
}
function locEnd(node) {
    return node.end;
}
exports.languages = [
    {
        name: 'svelte',
        parsers: ['svelte'],
        extensions: ['.svelte'],
        vscodeLanguageIds: ['svelte'],
    },
];
exports.parsers = {
    svelte: {
        hasPragma: print_1.hasPragma,
        parse: async (text, options) => {
            try {
                let _parse = compiler_1.parse;
                if (options.svelte5CompilerPath) {
                    try {
                        _parse = (await Promise.resolve(`${options.svelte5CompilerPath}`).then(s => require(s))).parse;
                    }
                    catch (e) {
                        console.warn(`Failed to load Svelte 5 compiler from ${options.svelte5CompilerPath}`);
                        console.warn(e);
                        options.svelte5CompilerPath = undefined;
                    }
                }
                const root = _parse(text);
                root.__isRoot = true;
                // TODO this will need to be done once we switch to the modern parser output:
                // Prettier does a sanity check on ast.comments after printing
                // to verify all comments were printed. Since the comments array
                // includes script/style comments already handled by embedded
                // parsers, we stash the full array on _comments and remove
                // comments so Prettier doesn't try to process them itself.
                // We then manually attach attribute comments in embed().
                // (root as ASTNode)._comments = root.comments;
                // delete root.comments;
                return root;
            }
            catch (err) {
                if (err.start != null && err.end != null) {
                    // Prettier expects error objects to have loc.start and loc.end fields.
                    // Svelte uses start and end directly on the error.
                    err.loc = {
                        start: err.start,
                        end: err.end,
                    };
                }
                throw err;
            }
        },
        preprocess: (text, options) => {
            const result = (0, snipTagContent_1.snipScriptAndStyleTagContent)(text);
            text = result.text.trim();
            // Prettier sets the preprocessed text as the originalText in case
            // the Svelte formatter is called directly. In case it's called
            // as an embedded parser (for example when there's a Svelte code block
            // inside markdown), the originalText is not updated after preprocessing.
            // Therefore we do it ourselves here.
            options.originalText = text;
            options._svelte_ts = result.isTypescript;
            return text;
        },
        locStart,
        locEnd,
        astFormat: 'svelte-ast',
    },
    svelteExpressionParser: {
        ...babelParser,
        parse: (text, options) => {
            const ast = babelParser.parse(text, options);
            let program = ast.program.body[0];
            if (!options._svelte_asFunction) {
                program = program.expression;
            }
            return { ...ast, program };
        },
    },
    svelteTSExpressionParser: {
        ...typescriptParser,
        parse: (text, options) => {
            const ast = typescriptParser.parse(text, options);
            let program = ast.program.body[0];
            if (!options._svelte_asFunction) {
                program = program.expression;
            }
            return { ...ast, program };
        },
    },
};
exports.printers = {
    'svelte-ast': {
        print: print_1.print,
        embed: embed_1.embed,
        // @ts-expect-error Prettier's type definitions don't include getVisitorKeys
        getVisitorKeys: embed_1.getVisitorKeys,
        isBlockComment(comment) {
            return comment.type === 'Block';
        },
        printComment(commentPath) {
            const comment = commentPath.getValue();
            if (comment.type === 'Line') {
                return '//' + comment.value.replace(/\r$/, '');
            }
            return '/*' + comment.value + '*/';
        },
    },
};
var options_1 = require("./options");
Object.defineProperty(exports, "options", { enumerable: true, get: function () { return options_1.options; } });
