// svelte-eslint-parser/index.js
const visitorKeys = {
  Program: ["body"],
};

const zeroLoc = {
  start: { line: 1, column: 0 },
  end: { line: 1, column: 0 },
};

export default {
  name: "svelte-eslint-parser",
  VisitorKeys: visitorKeys,
  parseForESLint(code, options = {}) {
    return {
      ast: {
        type: "Program",
        sourceType: "module",
        body: [],
        range: [0, code.length],
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: code.length },
        },
        comments: options.comment === true ? [{ type: "Block", range: [0, 0], loc: zeroLoc, value: "" }] : [],
        tokens: options.tokens === true ? [{ type: "Identifier", range: [0, 0], loc: zeroLoc, value: "" }] : [],
      },
      visitorKeys,
      services: {
        commentFlag: options.comment === true,
        tokensFlag: options.tokens === true,
        locFlag: options.loc === true,
        rangeFlag: options.range === true,
        rawFlag: options.raw === true,
      },
    };
  },
};
