// svelte-eslint-parser/index.js
const visitorKeys = {
  Program: ["body"],
};

function getLineColumn(code, index) {
  let line = 1;
  let lineStart = 0;

  for (let i = 0; i < index; i++) {
    if (code.charCodeAt(i) === 10) {
      line += 1;
      lineStart = i + 1;
    }
  }

  return { line, column: index - lineStart };
}

function getLoc(code, start, end) {
  return {
    start: getLineColumn(code, start),
    end: getLineColumn(code, end),
  };
}

export default {
  name: "svelte-eslint-parser",
  VisitorKeys: visitorKeys,
  parseForESLint(code, options = {}) {
    const commentEnd = code.indexOf("-->") + 3;
    const h1Start = code.indexOf("<h1");
    const program = {
      type: "Program",
      sourceType: "module",
      body: [],
      range: [0, code.length],
      loc: getLoc(code, 0, code.length),
      comments:
        options.comment === true && commentEnd >= 3
          ? [
              {
                type: "Block",
                value: " marker ",
                range: [0, commentEnd],
                loc: getLoc(code, 0, commentEnd),
              },
            ]
          : [],
      tokens:
        options.tokens === true && h1Start >= 0
          ? [
              {
                type: "Punctuator",
                value: "<",
                range: [h1Start, h1Start + 1],
                loc: getLoc(code, h1Start, h1Start + 1),
              },
              {
                type: "JSXIdentifier",
                value: "h1",
                range: [h1Start + 1, h1Start + 3],
                loc: getLoc(code, h1Start + 1, h1Start + 3),
              },
            ]
          : [],
    };

    return {
      ast: program,
      visitorKeys,
      services: {
        locFlag: options.loc === true,
        rangeFlag: options.range === true,
        rawFlag: options.raw === true,
        commentFlag: options.comment === true,
        tokenFlag: options.tokens === true,
      },
    };
  },
};
