// eslint-plugin-svelte/index.js
const plugin = {
  meta: {
    name: "svelte",
  },
  rules: {
    "ast-metadata-flags": {
      create(context) {
        return {
          Program(node) {
            const services = context.sourceCode.parserServices;
            const comments = context.sourceCode.getAllComments();
            const tokens = context.sourceCode.getTokens(node);

            context.report({
              node,
              message: [
                `locFlag: ${services.locFlag === true}`,
                `rangeFlag: ${services.rangeFlag === true}`,
                `rawFlag: ${services.rawFlag === true}`,
                `commentFlag: ${services.commentFlag === true}`,
                `tokenFlag: ${services.tokenFlag === true}`,
                `commentCount: ${comments.length === 1}`,
                `tokenCount: ${tokens.length === 2}`,
              ].join("; "),
            });
          },
        };
      },
    },
  },
  configs: {
    recommended: {
      jsPlugins: ["eslint-plugin-svelte"],
      rules: {
        "svelte/ast-metadata-flags": "error",
      },
    },
  },
};

export default plugin;
