// svelte-eslint-parser/index.js
const visitorKeys = {
  Program: ["body"],
  SvelteElement: ["attributes", "children"],
  SvelteClassName: [],
  SvelteText: [],
};

export default {
  name: "svelte-eslint-parser",
  VisitorKeys: visitorKeys,
  parseForESLint(code, options = {}) {
    const classStart = code.indexOf("greeting");
    const textStart = code.indexOf("Hello");

    const classNode = {
      type: "SvelteClassName",
      value: "greeting",
      range: [classStart, classStart + "greeting".length],
    };
    const textNode = {
      type: "SvelteText",
      value: "Hello",
      range: [textStart, textStart + "Hello".length],
    };

    return {
      ast: {
        type: "Program",
        sourceType: "module",
        body: [
          {
            type: "SvelteElement",
            name: "h1",
            range: [0, code.length],
            attributes: [classNode],
            children: [textNode],
          },
        ],
        range: [0, code.length],
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: code.length },
        },
        comments: [],
        tokens: [],
      },
      visitorKeys,
      services: {
        isSvelte: true,
        filePath: options.filePath ?? null,
      },
    };
  },
};
