// eslint-plugin-svelte/index.js
const plugin = {
  meta: {
    name: "svelte",
  },
  rules: {
    "whole-file-edits": {
      meta: {
        fixable: "code",
        hasSuggestions: true,
      },
      create(context) {
        const hasParserServices = context.sourceCode.parserServices.isSvelte === true;
        return {
          SvelteClassName(node) {
            context.report({
              node,
              message: `fix class attribute; parserServices: ${hasParserServices}`,
              fix(fixer) {
                return fixer.replaceText(node, "welcome");
              },
            });
          },
          SvelteText(node) {
            context.report({
              node,
              message: `suggest greeting text; parserServices: ${hasParserServices}`,
              suggest: [
                {
                  desc: "Use Hi",
                  fix(fixer) {
                    return fixer.replaceText(node, "Hi");
                  },
                },
                {
                  desc: "Use Hey",
                  fix(fixer) {
                    return fixer.replaceText(node, "Hey");
                  },
                },
              ],
            });
          },
        };
      },
    },
  },
  configs: {
    recommended: {
      jsPlugins: ["eslint-plugin-svelte"],
      rules: {
        "svelte/whole-file-edits": "error",
      },
    },
  },
};

export default plugin;
