diff --git a/dist/theme/github-dark.js b/dist/theme/github-dark.js
index 3758dd94146e9434a29d835c59c8e8040544da10..3faefcceab2e060b4020b722dbe1c59bc9ea2323 100644
--- a/dist/theme/github-dark.js
+++ b/dist/theme/github-dark.js
@@ -46,7 +46,7 @@ const githubDarkTheme = EditorView.theme({
     '.cm-searchMatch.cm-searchMatch-selected': {
         backgroundColor: config.selection
     },
-    '.cm-activeLine': { backgroundColor: config.activeLine },
+    '.cm-activeLine': { backgroundColor: 'transparent' },
     '.cm-selectionMatch': { backgroundColor: config.selection },
     '&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket': {
         backgroundColor: config.matchingBracket,