// eslint-plugin-svelte/index.js
const plugin = {
  meta: {
    name: "svelte",
  },
  rules: {
    "baseline-flags": {
      create(context) {
        return {
          Program(node) {
            const services = context.sourceCode.parserServices;
            const sourceAst = context.sourceCode.ast;
            context.report({
              node,
              message: [
                `commentFlag: ${services.commentFlag === true}`,
                `tokensFlag: ${services.tokensFlag === true}`,
                `locFlag: ${services.locFlag === true}`,
                `rangeFlag: ${services.rangeFlag === true}`,
                `rawFlag: ${services.rawFlag === true}`,
                `commentsApi: ${Array.isArray(sourceAst.comments) && sourceAst.comments.length === 1}`,
                `tokensApi: ${Array.isArray(sourceAst.tokens) && sourceAst.tokens.length === 1}`,
                `wholeFile: ${context.sourceCode.text.includes('<h1 class="greeting">Hello {name}</h1>')}`,
              ].join('; '),
            });
          },
        };
      },
    },
  },
  configs: {
    recommended: {
      jsPlugins: ["eslint-plugin-svelte"],
      rules: {
        "svelte/baseline-flags": "error",
      },
    },
  },
};

export default plugin;
