// svelte/compiler.js
"use strict";

/**
 * Minimal CommonJS `svelte/compiler` fixture for Oxfmt's real `prettier-plugin-svelte` tests.
 *
 * It intentionally supports only the narrow document shapes used in this fixture directory:
 * - one top-level `<script>` block
 * - one top-level `<style>` block
 * - simple top-level markup with a single element containing text and `{identifier}` tags
 */

function parse(source) {
  const script = findTopLevelTag(source, "script");
  const style = findTopLevelTag(source, "style");
  const occupiedRanges = [script, style]
    .filter(Boolean)
    .map((tag) => [tag.start, tag.end])
    .sort((a, b) => a[0] - b[0]);

  const htmlStart = findMarkupStart(source, occupiedRanges);
  const htmlEnd = findMarkupEnd(source, occupiedRanges);
  const html = parseFragment(source, htmlStart, htmlEnd);

  html.start = 0;
  html.end = source.length;

  return {
    type: "Root",
    start: 0,
    end: source.length,
    html,
    instance: script ? createTopLevelTagNode("Script", source, script) : null,
    module: null,
    css: style ? createTopLevelTagNode("Style", source, style) : null,
    js: null,
    comments: [],
  };
}

function findTopLevelTag(source, tagName) {
  const pattern = new RegExp(`<${tagName}\\b([^>]*)>([\\s\\S]*?)<\\/${tagName}>`, "i");
  const match = pattern.exec(source);
  if (!match || match.index == null) return null;

  const fullText = match[0];
  const start = match.index;
  const end = start + fullText.length;
  const openEnd = start + fullText.indexOf(">") + 1;
  const closeTagText = `</${tagName}>`;
  const contentEnd = end - closeTagText.length;

  return {
    tagName,
    start,
    end,
    openEnd,
    contentStart: openEnd,
    contentEnd,
    attributesText: match[1] ?? "",
  };
}

function createTopLevelTagNode(type, source, tag) {
  return {
    type,
    start: tag.start,
    end: tag.end,
    attributes: parseAttributes(source, tag),
    children: [],
    comments: [],
    content: {
      start: tag.contentStart,
      end: tag.contentEnd,
      children: [],
    },
  };
}

function parseAttributes(source, tag) {
  const attributesText = tag.attributesText;
  if (!attributesText) return [];

  const attrs = [];
  const tagHead = source.slice(tag.start, tag.openEnd);
  const regex = /([^\s=]+)(?:=(?:"([^"]*)"|'([^']*)'|([^\s>]+)))?/g;
  let searchFrom = 0;
  let match;

  while ((match = regex.exec(attributesText)) !== null) {
    const [fullMatch, name, dq, sq, bare] = match;
    const value = dq ?? sq ?? bare ?? null;
    const attrStartInHead = tagHead.indexOf(fullMatch, searchFrom);
    const start = attrStartInHead === -1 ? tag.start : tag.start + attrStartInHead;
    const end = start + fullMatch.length;
    searchFrom = Math.max(attrStartInHead + fullMatch.length, searchFrom);

    attrs.push({
      type: "Attribute",
      name,
      start,
      end,
      value: value === null
        ? true
        : [{
            type: "Text",
            start,
            end,
            data: value,
            raw: value,
          }],
    });
  }

  return attrs;
}

function findMarkupStart(source, occupiedRanges) {
  let cursor = 0;
  for (const [start, end] of occupiedRanges) {
    const gap = source.slice(cursor, start);
    if (gap.trim().length > 0) return cursor + firstNonWhitespace(gap);
    cursor = end;
  }

  const remainder = source.slice(cursor);
  return remainder.trim().length === 0 ? cursor : cursor + firstNonWhitespace(remainder);
}

function findMarkupEnd(source, occupiedRanges) {
  let cursor = source.length;
  for (let i = occupiedRanges.length - 1; i >= 0; i--) {
    const [start, end] = occupiedRanges[i];
    const gap = source.slice(end, cursor);
    if (gap.trim().length > 0) return end + lastNonWhitespaceOffset(gap);
    cursor = start;
  }

  const prefix = source.slice(0, cursor);
  return prefix.trim().length === 0 ? cursor : lastNonWhitespaceOffset(prefix);
}

function firstNonWhitespace(text) {
  const match = /\S/.exec(text);
  return match ? match.index : 0;
}

function lastNonWhitespaceOffset(text) {
  for (let index = text.length - 1; index >= 0; index--) {
    if (!/\s/.test(text[index])) return index + 1;
  }
  return 0;
}

function parseFragment(source, start, end) {
  if (start >= end) {
    return { type: "Fragment", start, end, children: [] };
  }

  const markup = source.slice(start, end);
  const elementMatch = /^<([A-Za-z0-9:-]+)([^>]*)>([\s\S]*)<\/\1>$/.exec(markup.trim());
  if (!elementMatch) {
    return {
      type: "Fragment",
      start,
      end,
      children: [{
        type: "Text",
        start,
        end,
        data: markup,
        raw: markup,
      }],
    };
  }

  const fullElementText = elementMatch[0];
  const elementStart = source.indexOf(fullElementText, start);
  const elementEnd = elementStart + fullElementText.length;
  const openTagEnd = source.indexOf(">", elementStart) + 1;
  const childrenStart = openTagEnd;
  const childrenEnd = elementEnd - (`</${elementMatch[1]}>`.length);

  return {
    type: "Fragment",
    start: elementStart,
    end: elementEnd,
    children: [createElementNode(elementMatch[1], elementStart, elementEnd, source, childrenStart, childrenEnd)],
  };
}

function createElementNode(name, start, end, source, childrenStart, childrenEnd) {
  return {
    type: "Element",
    name,
    start,
    end,
    attributes: [],
    children: parseElementChildren(source, childrenStart, childrenEnd),
  };
}

function parseElementChildren(source, start, end) {
  const text = source.slice(start, end);
  const children = [];
  const mustacheRegex = /\{([A-Za-z_$][\w$]*)\}/g;
  let lastIndex = 0;
  let match;

  while ((match = mustacheRegex.exec(text)) !== null) {
    const matchStart = start + match.index;
    const matchEnd = matchStart + match[0].length;

    if (match.index > lastIndex) {
      const rawText = text.slice(lastIndex, match.index);
      children.push({
        type: "Text",
        start: start + lastIndex,
        end: start + match.index,
        data: rawText,
        raw: rawText,
      });
    }

    children.push({
      type: "MustacheTag",
      start: matchStart,
      end: matchEnd,
      expression: {
        type: "Identifier",
        start: matchStart + 1,
        end: matchEnd - 1,
        name: match[1],
      },
    });

    lastIndex = match.index + match[0].length;
  }

  if (lastIndex < text.length) {
    const rawText = text.slice(lastIndex);
    children.push({
      type: "Text",
      start: start + lastIndex,
      end,
      data: rawText,
      raw: rawText,
    });
  }

  return children;
}

module.exports = {
  parse,
};
