// prettier-plugin-svelte/src/print/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isASTNode = isASTNode;
exports.isPreTagContent = isPreTagContent;
exports.flatten = flatten;
exports.findLastIndex = findLastIndex;
exports.replaceEndOfLineWith = replaceEndOfLineWith;
exports.getAttributeLine = getAttributeLine;
exports.printWithPrependedAttributeLine = printWithPrependedAttributeLine;
const prettier_1 = require("prettier");
const elements_1 = require("../lib/elements");
const snipTagContent_1 = require("../lib/snipTagContent");
/**
 * Determines whether or not given node
 * is the root of the Svelte AST.
 */
function isASTNode(n) {
    return n && n.__isRoot;
}
function isPreTagContent(path) {
    const stack = path.stack;
    return stack.some((node) => (node.type === 'Element' && node.name.toLowerCase() === 'pre') ||
        (node.type === 'Attribute' && !elements_1.formattableAttributes.includes(node.name)));
}
function flatten(arrays) {
    return [].concat.apply([], arrays);
}
function findLastIndex(isMatch, items) {
    for (let i = items.length - 1; i >= 0; i--) {
        if (isMatch(items[i], i)) {
            return i;
        }
    }
    return -1;
}
function replaceEndOfLineWith(text, replacement) {
    const parts = [];
    for (const part of text.split('\n')) {
        if (parts.length > 0) {
            parts.push(replacement);
        }
        if (part.endsWith('\r')) {
            parts.push(part.slice(0, -1));
        }
        else {
            parts.push(part);
        }
    }
    return parts;
}
function getAttributeLine(node, options) {
    const { hardline, line } = prettier_1.doc.builders;
    const hasThisBinding = (node.type === 'InlineComponent' && !!node.expression) ||
        (node.type === 'Element' && !!node.tag);
    const attributes = node.attributes.filter((attribute) => attribute.name !== snipTagContent_1.snippedTagContentAttribute);
    return options.singleAttributePerLine &&
        (attributes.length > 1 || (attributes.length && hasThisBinding))
        ? hardline
        : line;
}
function printWithPrependedAttributeLine(node, options, print) {
    return (path) => path.getNode().name !== snipTagContent_1.snippedTagContentAttribute
        ? [getAttributeLine(node, options), path.call(print)]
        : '';
}
